src/dbnode/storage/shard.go

```diff
@@ -2538,6 +2546,70 @@ func (s *dbShard) Repair(
 	return repairer.Repair(ctx, nsCtx, nsMeta, tr, s)
 }
 
+func (s *dbShard) AggregateTiles(
+	ctx context.Context,
+	reader fs.DataFileSetReader,
+	sourceNsID ident.ID,
+	sourceShard databaseShard,
+	opts AggregateTilesOptions,
+	wOpts series.WriteOptions,
+) error {
+	latestSourceVolume, err := sourceShard.latestVolume(opts.Start)
+	if err != nil {
+		return err
+	}
+
+	openOpts := fs.DataReaderOpenOptions{
+		Identifier: fs.FileSetFileIdentifier{
+			Namespace:   sourceNsID,
+			Shard:       sourceShard.ID(),
+			BlockStart:  opts.Start,
+			VolumeIndex: latestSourceVolume,
+		},
+		FileSetType: persist.FileSetFlushType,
+		//TODO add after https://github.com/chronosphereio/m3/pull/10 for proper streaming - OrderByIndex: true
+	}
+	if err := reader.Open(openOpts); err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	encodingOpts := encoding.NewOptions().SetBytesPool(s.opts.BytesPool())
+	bytesReader := bytes.NewReader(nil)
+	dataPointIter := m3tsz.NewReaderIterator(bytesReader, true, encodingOpts)
+	var lastWriteError error
+
+	for {
+		id, tags, data, _, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		data.IncRef()
+		bytesReader.Reset(data.Bytes())
+		dataPointIter.Reset(bytesReader, nil)
+
+		for dataPointIter.Next() {
+			dp, unit, annot := dataPointIter.Current()
+			_, err = s.writeAndIndex(ctx, id, tags, dp.Timestamp, dp.Value, unit, annot, wOpts, true)
```

Comment:
Is there some way to batch these writes?

Also, when doing a large aggregation, wouldn't we starve the CPU and block regular writes? How are we going to manage resource allocation?

Reply:
@gediminasgu